File: lua/examples/guards/scripts/Bullet.lua
---@uuid 12371d19-9f1a-4286-8486-add4ebaadaec
---@class Bullet : Script
---@field velocity Vector3
---@field remaining_sec number
---@field author_collider Node
Bullet = script_class()

Bullet.HitMessage = "BulletHit"

---@class BulletSeed
---@field prefab Prefab
---@field origin Vector3
---@field direction Vector3
---@field initial_velocity number
---@field author_collider Node
---@field range number
BulletSeed = nil

---@param seed BulletSeed
function Bullet:spawn(seed)
    local orientation = Quaternion:face_towards(seed.direction, Vector3.Y);
    local bullet = seed.prefab:instantiate_at(seed.origin, orientation);
    local script = bullet:find_script("Bullet")
    script.velocity = seed.direction:normalize():mul(seed.initial_velocity)
    script.remaining_sec = 1--seed.range / seed.initial_velocity
    script.author_collider = seed.author_collider
end

function Bullet:on_update(dt)
    self.remaining_sec = self.remaining_sec - dt
    if self.remaining_sec <= 0.0 then
        self.node:destroy()
        return;
    end
    local new_pos = self.node.local_position:add(self.velocity:mul(dt))

    ---@type RayCastOptions
    local opts = {
        ray_origin = self.node.local_position,
        ray_direction = self.velocity:normalize(),
        max_len = self.velocity:magnitude() * dt,
        sort_results = true
    }
    ---@type Intersection[]
    local results = Physics:cast_ray(opts)

    for i, hit in ipairs(results) do
        if hit.collider ~= self.author_collider then
            hit.collider:send_hierarchical(RoutingStrategy.Up, Bullet.HitMessage)
            self.node:destroy()
            return;
        end
    end

    self.node.local_position = new_pos;
end

File: cs/examples/Guards/Bullet.cs
using System;
using System.Collections.Generic;
using FyroxLite;

[Uuid("12371d19-9f1a-4286-8486-add4ebaadaec")]
public class Bullet : NodeScript
{
    private Vector3 velocity;
    private float remaining_sec;
    private Node author_collider;
    private float fraction;

    public struct BulletSeed
    {
        public Prefab Prefab;
        public Vector3 Origin;
        public Vector3 Direction;
        public float InitialVelocity;
        public Node AuthorCollider;
        public float Range;
        public int Fraction;
    }

    public static void Spawn(BulletSeed seed)
    {
        var orientation = Basis.LookingAt(seed.Direction, Vector3.Up).GetRotationQuaternion();
        var bullet = seed.Prefab.InstantiateAt(seed.Origin, orientation);
        var script = bullet.FindScript<Bullet>();
        script.velocity = seed.Direction.Normalized() * seed.InitialVelocity;
        script.remaining_sec = seed.Range / seed.InitialVelocity;
        script.author_collider = seed.AuthorCollider;
        script.fraction = seed.Fraction;
    }

    protected override void OnUpdate(float deltaTime)
    {
        remaining_sec -= deltaTime;
        if (remaining_sec <= 0.0f)
        {
            Node.Destroy();
            return;
        }

        List<Intersection> results = Physics.CastRay(new RayCastOptions
        {
            RayOrigin = Node.LocalPosition,
            RayDirection = velocity.Normalized(),
            MaxLen = velocity.Length() * deltaTime,
            SortResults = true
        });

        foreach (var hit in results)
        {
            if (hit.Collider != author_collider)
            {
                // scene from the Lua version of game is used, and Lua stores any number as f32
                var fraction = (int)this.fraction;

                hit.Collider.SendHierarchical(RoutingStrategy.Up, new BulletHitMessage { Fraction = fraction });
                Node.Destroy();
                return;
            }
        }

        Node.LocalPosition += velocity * deltaTime;
    }
}